graspologic/embed/n2v.py

```diff
@@ -0,0 +1,402 @@
+# Copyright (c) Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import logging
+import math
+import random
+import time
+from typing import Any, List, Optional, Tuple
+
+import networkx as nx
+import numpy as np
+
+
+def node2vec_embed(
+    graph: nx.Graph,
+    num_walks: int = 10,
+    walk_length: int = 80,
+    return_hyperparameter: int = 1,
+    inout_hyperparameter: int = 1,
+    dimensions: int = 128,
+    window_size: int = 10,
+    workers: int = 8,
+    iterations: int = 1,
+    interpolate_walk_lengths_by_node_degree: bool = True,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Generates a node2vec embedding from a given graph. Will follow the word2vec algorithm to create the embedding.
+
+    Parameters
+    ----------
+
+    graph: nx.Graph
+        A networkx graph. If the graph is unweighted, the weight of each edge will default to 1
```

Comment:
not-holding-this-PR-up comment: I thought node2vec would work fine on a directed graph?

Reply:
ah i see you actually check as we discussed - then my comment is, should this be Union[nx.Graph, nx.DiGraph]?

Reply:
I'm not entirely sure - mostly because DiGraph is a subclass of Graph.  That's why you can't do `isinstance(graph, nx.Graph)` to tell if something is undirected, because a `DiGraph` would make that be true too - you have to use `is_directed()` on it instead.

Type hinting has that annoying thing where it kinda fits in as a documentation system almost more so than a type system, so the normal "type system" ways of describing a parameter may not be as applicable (and thus, Union[nx.Graph, nx.DiGraph] would be more appropriate).  I... really am not sure, hah!

Reply:
The preconditions are also going to have to kick out the Multigraph variants, so it probably makes sense to explicitly say ONLY those 2.  It's... not what you normally do in a type system, but type hinting isn't a type system.  it's so... weird.